File: 15-iOS 雷达动画图片方式/iOS 雷达动画/ViewController.m
//
//  ViewController.m
//  iOS 雷达动画
//
//  Created by Xinxibin on 16/7/22.
//  Copyright © 2016年 GloryMan. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property (nonatomic,strong) CAShapeLayer * shapeLineLayer;

@property (weak, nonatomic) IBOutlet UIImageView *imageView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
//    self.imageView.center = CGPointMake(self.view.center.x , self.view.center.y + 175 + 100);

}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    

    [self setUI];
    [self startImageView];
}

- (void)startImageView {
    
    
    CABasicAnimation * animation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    animation.duration = 5;
    animation.toValue = [NSNumber numberWithFloat:M_PI * 2.0];
    animation.repeatCount = HUGE;
    animation.cumulative = true;
    
    [self.imageView.layer addAnimation:animation forKey:@"rotationAnimation"];
    
}

- (void)setUI {
    
    // 绘制四个圆圈
    CAShapeLayer * shapeLayer = [CAShapeLayer layer];
    UIBezierPath * path = [UIBezierPath bezierPath];
    [path addArcWithCenter:self.view.center radius:25 startAngle:0 endAngle:2 * M_PI clockwise:true];
    
    UIBezierPath * path2 = [UIBezierPath bezierPath];
    [path2 addArcWithCenter:self.view.center radius:75 startAngle:0 endAngle:2 * M_PI clockwise:true];
    [path appendPath:path2];
    
    UIBezierPath * path3 = [UIBezierPath bezierPath];
    [path3 addArcWithCenter:self.view.center radius:125 startAngle:0 endAngle:2 * M_PI clockwise:true];
    [path appendPath:path3];
    
    UIBezierPath * path4 = [UIBezierPath bezierPath];
    [path4 addArcWithCenter:self.view.center radius:175 startAngle:0 endAngle:2 * M_PI clockwise:true];
    [path appendPath:path4];
    
    shapeLayer.path = path.CGPath;
    shapeLayer.strokeColor = [UIColor blackColor].CGColor;
    shapeLayer.fillColor   = [UIColor clearColor].CGColor;
    shapeLayer.opacity = 0.5;
    [self.view.layer addSublayer:shapeLayer];
    
    // 绘制一条线
    UIBezierPath * linePath = [UIBezierPath bezierPath];
    [linePath moveToPoint:CGPointMake(12.5, self.view.center.y)];
    [linePath addLineToPoint:CGPointMake(self.view.center.x, self.view.center.y)];
    
    self.shapeLineLayer = [CAShapeLayer layer];
    self.shapeLineLayer.frame = self.view.bounds;
    self.shapeLineLayer.lineWidth = 1;
    self.shapeLineLayer.path = linePath.CGPath;
    self.shapeLineLayer.strokeColor = [UIColor redColor].CGColor;

    [self.view.layer addSublayer:self.shapeLineLayer];
    
    
    [self startAnimationLine];
    
}

- (void)startAnimationLine {
    
    CABasicAnimation * animation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    animation.fromValue = [NSNumber numberWithFloat:0.0];
    animation.toValue   = [NSNumber numberWithFloat:2 * M_PI];
    
    CAAnimationGroup * groupAnimation = [CAAnimationGroup animation];
    groupAnimation.duration = 5;
    groupAnimation.repeatCount = HUGE;
    groupAnimation.animations = @[animation];
    
    [self.shapeLineLayer addAnimation:groupAnimation forKey:@""];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: 1-ios 渐变loading/ios 渐变loading/ViewController.swift
//
//  ViewController.swift
//  ios 渐变loading
//
//  Created by Xinxibin on 16/8/3.
//  Copyright © 2016年 GloryMan. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        createLine()
    }

    func createLine() {
        
        let path = UIBezierPath()
        path.moveToPoint(CGPointMake(20,100))
        path.addLineToPoint(CGPointMake(40, 100))
        
        let shapeLayer              = CAShapeLayer()
        shapeLayer.lineWidth        = 3
        shapeLayer.strokeColor      = UIColor.blackColor().CGColor
        shapeLayer.fillColor        = UIColor.clearColor().CGColor
        shapeLayer.path             = path.CGPath
        
        self.view.layer.addSublayer(shapeLayer);
        
        // 基础渐变
        // 所以左上角坐标是{0, 0}，右下角坐标是{1, 1}
        let gradientLayer           = CAGradientLayer()
        gradientLayer.frame         = CGRect(x: 10, y: 120, width: 355, height: 20)
        self.view.layer.addSublayer(gradientLayer)
        gradientLayer.colors        = [ UIColor.blackColor().CGColor, UIColor.whiteColor().CGColor ]
        gradientLayer.startPoint    = CGPointMake(0, 0.5)
        gradientLayer.endPoint      = CGPointMake(1, 0.5)
        
        
        // 多重渐变
        /**
         colors属性可以包含很多颜色，所以创建一个彩虹一样的多重渐变也是很简单的
         默认情况下，这些颜色在空间上均匀地被渲染，但是我们可以用locations属性来调整空间。
         locations属性是一个浮点数值的数组（以NSNumber包装
         单位坐标系进行标定。0.0代表着渐变的开始，1.0代表着结束。
         */
        let gradientLayer1           = CAGradientLayer()
        gradientLayer1.frame         = CGRect(x: 10, y: 160, width: 355, height: 20)
        self.view.layer.addSublayer(gradientLayer1)
        gradientLayer1.colors        = [ UIColor.redColor().CGColor, UIColor.yellowColor().CGColor,UIColor.greenColor().CGColor ]
        gradientLayer1.startPoint    = CGPointMake(0, 0.5)
        gradientLayer1.endPoint      = CGPointMake(1, 0.5)
        gradientLayer1.locations     = [0.3,0.7,0.8]
        
        
        // 圆形渐变进度条
        let path2       = UIBezierPath(arcCenter: CGPointMake(375.0/2, 667.0/2), radius: 50, startAngle: 0, endAngle: 2 * CGFloat(M_PI), clockwise: true)
        let shapeLayer2     = CAShapeLayer()
        shapeLayer2.frame   = CGRectMake(0, 0, 100, 100)
        shapeLayer2.strokeColor = UIColor.blackColor().CGColor
        shapeLayer2.fillColor   = UIColor.clearColor().CGColor
        shapeLayer2.lineCap     = kCALineCapRound
        shapeLayer2.lineWidth   = 4
        shapeLayer2.opacity     = 0.25
        shapeLayer2.path        = path2.CGPath
        self.view.layer.addSublayer(shapeLayer2)
    }

}







